common/persistence/wrappers/templates/metered_execution.tmpl

```diff
@@ -0,0 +1,75 @@
+import (
+	"context"
+
+	"github.com/uber/cadence/common/config"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
+)
+
+{{ $decorator := (printf "metered%s" .Interface.Name) }}
+{{ $interfaceName := .Interface.Name }}
+
+// {{$decorator}} implements {{.Interface.Type}} interface instrumented with rate limiter.
+type {{$decorator}} struct {
+    base
+    wrapped     {{.Interface.Type}}
+}
+
+// New{{.Interface.Name}} creates a new instance of {{.Interface.Name}} with ratelimiter.
+func New{{.Interface.Name}}(
+	wrapped      persistence.{{.Interface.Name}},
+	metricClient metrics.Client,
+	logger       log.Logger,
+	cfg          *config.Persistence,
+    sampleLoggingRate dynamicconfig.IntPropertyFn,
+    enableShardIDMetrics dynamicconfig.BoolPropertyFn,
+) persistence.{{.Interface.Name}} {
+    return &{{$decorator}}{
+        wrapped: wrapped,
+        base:       base{
+			metricClient:                  metricClient,
+			logger:                        logger.WithTags(tag.ShardID(wrapped.GetShardID())),
+			enableLatencyHistogramMetrics: cfg.EnablePersistenceLatencyHistogramMetrics,
+			sampleLoggingRate:             sampleLoggingRate,
+			enableShardIDMetrics:          enableShardIDMetrics,
+        },
+    }
+}
+
+{{range $methodName, $method := .Interface.Methods}}
+    {{- if (and $method.AcceptsContext $method.ReturnsError)}}
+        func (c *{{$decorator}}) {{$method.Declaration}} {
+        	op := func() error {
+        	    {{$method.ResultsNames}} = c.wrapped.{{$method.Call}}
+        		{{ if ge (len $method.Results) 2 -}}
+        		    c.emptyMetric("{{$interfaceName}}.{{$methodName}}", {{(index $method.Params 1).Name}}, {{(index $method.Results 0).Name}}, err)
+        		{{ end -}}
+        		return err
+        	}
+
+        	{{$scopeName := printf "metrics.Persistence%sScope" $methodName}}
+
+            {{ if gt (len $method.Params) 1 -}}
+                {{ $reqName := (index $method.Params 1).Name }}
+                if domainName, hasDomainName := getDomainNameFromRequest({{$reqName}}); hasDomainName {
+                    logTags := append([]tag.Tag{tag.WorkflowDomainName(domainName)}, getCustomLogTags({{$reqName}})...)
+                    c.logger.SampleInfo("Persistence CreateWorkflowExecution called", c.sampleLoggingRate(), logTags...)
```

Comment:
I am not sure the log sample is correct?

Reply:
Good catch! Will fix.

Reply:
hmm I guess I need to adjust the test and make sure that logs are covered.

Reply:
Fixed and added tests